explorer/interpreter/type_checker.cpp

```diff
@@ -2335,10 +2335,20 @@ auto TypeChecker::MatchImpl(const InterfaceType& iface,
       *trace_stream_ << "matched with " << *impl.type << " as "
                      << *impl.interface << "\n\n";
     }
-    CARBON_ASSIGN_OR_RETURN(
-        const auto* subst_witness,
-        SubstituteCast<Witness>(*bindings_or_error, impl.witness));
-    return {subst_witness};
+
+    auto subst_witness_or_error =
+        SubstituteCast<Witness>(*bindings_or_error, impl.witness);
+
+    if (!subst_witness_or_error.ok()) {
+      return ErrorBuilder(subst_witness_or_error.error().location())
+             << subst_witness_or_error.error().message() << "\n"
+             << "NOTE: " << source_loc.ToString()
+             << ": in instantiation of `impl " << *impl.type << " as "
+             << GetName(impl.interface->declaration()) << "` with `"
+             << *impl.type << " = " << *impl_type << "` required here";
+    }
```

Comment:
I think this note should be built in `InstantiateImplDeclaration` instead of here. In principle, there could be other reasons that substitution could fail other than an instantiation error, and there could be other calls to `Substitute` that end up triggering template instantiation other than this one.

In order for that to work, I expect you'll need to change `Substitute` to take a source location and update all of its callers to pass one in. That's a fair amount of work, but it should be largely mechanical.

Reply:
I agree, my first implementation was actually in the `InstantiateImplDeclaration` function. But I realized that in addition to adding a `SourceLocation` argument to `Substitute,`  I would also need to modify the `SubstituteImpl` which internally calls one of the many `operator()` methods of `TransformBase` or one of its subclasses. 

Following this solution forces me to add a `SourceLocation` to every `operator()` method, creating a lot of unused code. Are there any other possible solutions to look into or should I just stick with this?

Reply:
Found an easy way by simply storing the source location in the `SubstituteTransform` class.

Reply:
In the implementation, you're currently defaulting the `SourceLocation` to unknown in most cases. However, as zygoloid noted, there could be other reasons that the substitution could fail. While the current approach fixes the one specific message, I think the earlier comment indicates the SourceLocation should be provided without defaulting.

Reply:
dd386fa4aea3792994dd816591a2fbd5618f7753 is my attempt at providing a SourceLocation to all the Substitute calls. I'm still not really sure about the location of the calls in method `ConstraintTypeBuilder::AddAndSubstitute`. I tried my best at inferring the correct SourceLocation for all the calls but it definitely required thorough review.